docs/learn/learn-xcm.md

```diff
@@ -0,0 +1,299 @@
+
+It’s an ultra-high level non-Turing-complete computer whose instructions are designed to be roughly at the same level as transactions. Messages are one or more XCM instructions. The program executes until it either runs to the end or hits an error, at which point it finishes up and halts. An XCM executor following the XCVM specification is provided by Parity, and it can be extended or customized, or even ignored altogether and users can create their own construct that follows the XCVM spec.
+
+A _message_ in XCM is simply just a programme that runs on the `XCVM`: in other words, one or more
+XCM instructions. To learn more about the XCVM and the XCM Format, see the latest
+[blog post](https://medium.com/polkadot-network/xcm-the-cross-consensus-message-format-3b77b1373392)
+by Dr. Gavin Wood.
+
+XCM Instructions might change a register, they might change the state of the consensus system or both.
+
+One example of such an instruction would be `TransferAsset` which is used to transfer an asset to some other address on the remote system. It needs to be told which asset(s) to transfer and to whom/where the asset is to be transferred.
+
+```
+enum Instruction {
+    TransferAsset {
+        assets: MultiAssets,
+        beneficiary: MultiLocation,
+    }
+    /* snip */
+}
+```
+A `MultiAsset` is a general identifier for an asset. It may represent both fungible and non-fungible assets, and in the case of a fungible asset, it represents some defined amount of the asset.
+
+A `MultiLocation` is a relative identifier, meaning that it can only be used to define the relative path between two locations, and cannot generally be used to refer to a location universally. Much like a relative file-system path will first begin with any "../" components used to ascend into to the containing directory, followed by the directory names into which to descend, a `MultiLocation` has two main parts to it: the number of times to ascend into the outer consensus from the local and then an interior location within that outer consensus.
+
+## Cross-Consensus Protocols (XCMP, VMP, HRMP)
+
+With the XCM format established, common patterns for protocols of these messages are needed.
+Polkadot implements two message passing protocols for acting on XCM messages between its constituent
+parachains.
+
+### XCMP (Cross-Chain Message Passing)
+
+:::caution
+
+XCMP is currently under development and most cross-chain message passing uses HRMP channels for the time being.
+
+:::
+
+_Cross-Chain Message Passing_ secure message passing between parachains. There are two variants:
+_Direct_ and _Relayed_.
+
+- With _Direct_, message data goes direct between parachains and is O(1) on the side of the
+  Relay-chain and is very scalable.
+- With _Relayed_, message data is passed via the Relay-chain, and piggy-backs over VMP. It is much
```

Comment:
I'm not 100% sure whether the HRMP pallet actually piggybacks over VMP, someone should check this.

Reply:
This section is XCMP specific, HRMP section is below